SimpleAdaptiveTauLeaping solver

sivasathyaseeelan opened the pull request:
Adds `SimpleAdaptiveTauLeaping`, which switches between an explicit and an implicit tau-leaping step according to whether the system currently looks stiff.

Explicit tau-leaping is efficient on non-stiff systems but is limited by the fastest reaction, while an implicit step lifts that limit at the cost of a nonlinear solve per step. This solver measures stiffness at each step and pays for the implicit step only when it is needed, following Cao et al. (2007).

```julia
using JumpProcesses

maj = MassActionJump([100.0, 100.0, 0.1], [[1 => 1], [2 => 1], [1 => 1]],
                     [[1 => -1, 2 => 1], [1 => 1, 2 => -1], [1 => -1, 3 => 1]])
prob = DiscreteProblem([100.0, 100.0, 0.0], (0.0, 1.0))
jprob = JumpProblem(prob, PureLeaping(), maj)

sol = solve(jprob, SimpleAdaptiveTauLeaping())                            # propensity-ratio test
sol = solve(jprob, SimpleAdaptiveTauLeaping(eigenvalue_check = true))     # eigenvalue test
```

Stiffness is judged from the spread of the propensities by default, or from the eigenvalue ratio of the drift Jacobian when `eigenvalue_check` is set. When a step is taken implicitly the tau-selection tolerance is relaxed by `implicit_epsilon_factor`, since that step is not restricted by the fast timescale it damps. The implicit step is available in fully implicit (`NewtonImplicitSolver`) and trapezoidal (`TrapezoidalImplicitSolver`) form.

## Notes on this revision

This has been rebased onto current `master`, which in the meantime merged `SimpleExplicitTauLeaping`. Step-size selection, mass action rate construction and `saveat` handling are shared with it rather than duplicated, and `validate_pure_leaping_inputs` is widened to cover both algorithms rather than gaining a second identical method.

Two problems from the earlier revision are fixed:

* **The eigenvalue-based stiffness check took eigenvalues of a non-square matrix.** It built the Jacobian of the propensities, which is `numjumps × numspecies`, and passed it straight to `eigvals`. That only works when a model happens to have as many reactions as species; an SEIR model (3 reactions, 4 species) raises `DimensionMismatch: matrix is not square`. The stiffness of the process is a property of the drift `dX/dt = Σⱼ νⱼ aⱼ(X)`, so the check now uses the square Jacobian `nu * da/dX`.
* **A non-convergent implicit step could spin forever.** The retry path halved `tau` and then `continue`d, but `tau` is recomputed at the top of each iteration, so the halved value was discarded. An explicit retry survives this because it redraws its Poisson counts, but the implicit solve is deterministic in `(u_current, tau)` and would fail identically on every pass. The bound is now carried across iterations, and a step that still fails at `dtmin` is reported as an error.

The algorithm type is also fully parameterized (`SimpleAdaptiveTauLeaping{T, S}`) rather than holding an abstractly-typed `solver` field.

## Testing

The existing SIR and SEIR mean-trajectory comparisons in `test/regular_jumps.jl` are extended to both formulations, checked against the exact SSA solution. A dedicated test covers the eigenvalue-based stiffness check on an SEIR model — a case with more species than reactions, which is exactly what the old Jacobian could not handle — and asserts the trajectory stays non-negative and conserves total population.

## Relationship to #500

This PR and #500 both introduce `AbstractImplicitSolver`, `NewtonImplicitSolver`, `TrapezoidalImplicitSolver` and the implicit-step machinery, since each has to stand alone against `master`. Those blocks are byte-identical between the two branches, so whichever merges first, rebasing the other is a clean deletion rather than a real conflict. Merging #500 first and then rebasing this one on top is the tidier order.

## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [X] All documentation related to code changes were updated
- [X] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API

sivasathyaseeelan committed:
SimpleAdaptiveTauLeaping

sivasathyaseeelan committed:
some

sivasathyaseeelan committed:
added jacobian stifness condition

sivasathyaseeelan committed:
refactor

sivasathyaseeelan committed:
refactor

sivasathyaseeelan committed:
refactor

sivasathyaseeelan committed:
test fix

sivasathyaseeelan committed:
bug fix

sivasathyaseeelan committed:
bug fix

sivasathyaseelan-ema committed:
Add a SimpleAdaptiveTauLeaping solver

Explicit tau-leaping is efficient on non-stiff systems but is limited by the
fastest reaction, while an implicit step lifts that limit at the cost of a
nonlinear solve. This adds a solver that measures stiffness at each step and
pays for the implicit step only when it is needed, following Cao et al. (2007).

Stiffness is judged from the spread of the propensities by default, or from the
eigenvalue ratio of the drift Jacobian when `eigenvalue_check` is set. When a
step is taken implicitly the tau-selection tolerance is relaxed by
`implicit_epsilon_factor`, since that step is not restricted by the fast
timescale it damps. The implicit step itself is available in fully implicit
(`NewtonImplicitSolver`) and trapezoidal (`TrapezoidalImplicitSolver`) form.

The eigenvalue test uses the Jacobian of the drift, `nu * da/dX`. The Jacobian
of the propensities alone is `numjumps x numspecies`, which is not square unless
a model happens to have as many reactions as species, so taking its eigenvalues
throws a DimensionMismatch on, for example, any SEIR model.

Step-size selection, mass action rate construction and `saveat` handling are
shared with `SimpleExplicitTauLeaping` rather than duplicated, and the existing
`validate_pure_leaping_inputs` method is widened to cover both algorithms.

A rejected implicit step halves `tau` and carries that bound across loop
iterations. An explicit retry redraws its Poisson counts and so eventually
succeeds at an unchanged `tau`, but the implicit solve is deterministic in
`(u_current, tau)` and would otherwise fail identically forever.

Tests extend the SIR and SEIR mean-trajectory comparisons to both implicit
formulations, and cover the eigenvalue-based stiffness check on a model with
more species than reactions.

ChrisRackauckas commented on src/simple_regular_solve.jl:
```diff
@@ -475,6 +571,290 @@ function DiffEqBase.solve(jump_prob::JumpProblem, alg::SimpleExplicitTauLeaping;
+        for j in 1:numjumps
+            for spec_idx in axes(nu, 1)
+                resid[spec_idx] -= nu[spec_idx, j] * rate_new[j] * tau
+            end
```
why does the adaptive explicit method have newton solves?

sivasathyaseeelan replied:
It should not, and it no longer does. This PR now defines no implicit machinery of its own.

You pointed at the copy of `implicit_equation!` that this branch carried, together with its own `AbstractImplicitSolver`, `NewtonImplicitSolver` and `TrapezoidalImplicitSolver` types. That duplication is removed. The branch is rebased on master, and the implicit step now comes from #643. A diff against master adds no `implicit_equation!`, no `solve_implicit` and none of those solver types.

On the name: this is not an explicit method. `SimpleExplicitTauLeaping` already adapts `tau`, so "adaptive" is the wrong word for what is new here. This algorithm decides at each step whether the system is stiff. If it is not stiff, it takes the explicit step. If it is stiff, it takes an implicit step, and that step is the algorithm given as `implicit_alg`, either `SimpleImplicitTauLeaping` or `SimpleTrapezoidalLeaping`. The Newton solve is therefore the one from #643, and only stiff steps pay for it. The method is Cao et al. (2007).

If you prefer a clearer name, I am happy to rename it. `SimpleAutoTauLeaping` would match the `Auto*` convention that OrdinaryDiffEq uses for stiffness switching, for example `AutoTsit5(Rosenbrock23())`. Tell me which you want.

The PR now adds these definitions only:

* `SimpleAdaptiveTauLeaping` and its keyword constructor
* `compute_tau_implicit`, which relaxes the tau tolerance for a stiff step
* `compute_drift_jacobian` and `is_stiff`, for the stiffness test
* the step loop and its `solve` method

ChrisRackauckas commented:
Rebase this onto master.

sivasathyaseeelan commented:
Done, rebased on master.

The rebase removes the duplicated implicit machinery. The adaptive solver now holds an `implicit_alg` field and calls the `solve_implicit` from #643, so the residual and the nonlinear solve are shared, not copied. The diff drops from 382 to 323 added lines, and it no longer touches `Project.toml`, because master already has `ADTypes` and `SimpleNonlinearSolve`.

Two bugs from the earlier revision are also fixed:

* The eigenvalue stiffness test called `eigvals` on the Jacobian of the propensities. That matrix is `numjumps x numspecies`, so it is not square unless a model has as many reactions as species, and an SEIR model raises `DimensionMismatch: matrix is not square`. The test now uses the square drift Jacobian `nu * da/dX`.
* A rejected implicit step halved `tau` and then continued, but `tau` was recomputed at the top of the loop, so the smaller value was lost. An explicit retry redraws its Poisson counts and recovers; the implicit solve is deterministic in `(u_current, tau)` and repeated the same failure. The loop now carries that bound, and a step that still fails at `dtmin` raises an error.
